### ISSUE TITLE
Null reference in ListStyles causes error when removing last item in a list

### ISSUE THREAD

**Issue by jrit** [ckeditor/ckeditor5#8642]
https://github.com/ckeditor/ckeditor5/blob/0ad1d6a0587cb8223f33e0d5b6d186ebcfb4b4a0/packages/ckeditor5-list/src/liststyleediting.js#L194

stack reads from bottom up here, this was after selecting a list item to paste onto

Cannot read property 'is' of null 
    CkEditor/node_modules/@ckeditor/ckeditor5-engine/src/model/position.js:931:14 Function._createAt
    CkEditor/node_modules/@ckeditor/ckeditor5-engine/src/model/model.js:635:23 nl.createPositionAt
    CkEditor/node_modules/@ckeditor/ckeditor5-engine/src/model/writer.js:607:20 _c.createPositionAt
    CkEditor/node_modules/@ckeditor/ckeditor5-list/src/liststyleediting.js:194:32 

Seems (without looking at it too much) the the referenced line needs to be updated to handle a null

**Comment by Reinmar** [comment]
Hi! Could you explain what are the steps to reproduce this issue?

**Comment by samfrjn11** [comment]
Experiencing the same issue.

```
ckeditorerror.js:64 Uncaught TypeError: Cannot read property 'is' of null
    at Function._createAt (position.js:931)
    at nl.createPositionAt (model.js:635)
    at fc.createPositionAt (writer.js:607)
    at liststyleediting.js:194
    at nl.change (model.js:178)
    at nl.listenTo.priority (liststyleediting.js:178)
    at nl.fire (emittermixin.js:221)
    at nl.<computed> [as deleteContent] (observablemixin.js:259)
    at Object.callback (deletecommand.js:113)
    at nl._runPendingChanges (model.js:830)
```

In my case I have
- Heading
-- paragraphs
-- list with 5 items
- Heading

When I select the last list item (triple click to select whole line) and hit "delete", it moves the Heading to the list, keeping the heading style in the list. At this point the above error is thrown.
![Schermafbeelding 2021-01-25 om 11 58 07](https://user-images.githubusercontent.com/13766544/105697273-ab194500-5f04-11eb-993d-fcf30a76016e.png)
![Schermafbeelding 2021-01-25 om 11 57 55](https://user-images.githubusercontent.com/13766544/105697277-ac4a7200-5f04-11eb-98ce-aeed4b4fa2bf.png)

**Comment by FilipTokarski** [comment]
I was trying to reproduce this error on the list [style docs demo](https://ckeditor.com/docs/ckeditor5/latest/features/lists/lists.html#demo-2) and also locally but could not.  
@jrit @samfrjn11 can you reproduce this on our docs demo or just locally? Without more information I'm not able to verify this.

**Comment by samfrjn11** [comment]
At first I had issues to reproduce myself. After messing around I had it again.
Then I spent quite a lot of time to see when and why it happens.

Exact steps how I reproduced (I did this on the [style docs demo](https://ckeditor.com/docs/ckeditor5/latest/features/lists/lists.html#demo-2)):

1. cmd+a and hit delete (to clear the demo content, I'm on MacOS)
2. Click "paragraph" and select Heading 1
3. type a title, I choose "my first heading"
4. hit enter to start a new paragraph
5. Click on bulleted list
6. write some bullet points
7. hit enter twice to start a new paragraph
8. click "paragraph", select "heading 1" and type another title. I choose "another heading"
9. Hit enter to go to the next paragraph
10. click "bulleted list" to create a new list
11. write some bullet points
12. now double click the last item of your bulleted list under your first heading
13. hit delete on your keyboard

I performed above steps as I typed it and it didn't work. i found out that the issue appears when your first list ENDS with 2 words (see screenshot). 

Edit: happens when more than one word is used.

So in short:
it happens when a list item has multiple words and is followed by a heading followed by another list.
![Schermafbeelding 2021-02-09 om 17 03 16](https://user-images.githubusercontent.com/13766544/107391539-12d0b200-6af9-11eb-9cd1-52f71af2a685.png)

**Comment by samfrjn11** [comment]
@FilipTokarski I'm going to make it even easier for you, paste following code in your ckeditor html element before executing the javascript:

`<h2>My first heading</h2><ul><li>with</li><li>some</li><li>bullet</li><li>points</li><li>of</li><li>my</li><li>first heading</li></ul><h2>Another title</h2><ul><li>with</li><li>some</li><li>other</li><li>bullet points</li></ul>`

**Comment by FilipTokarski** [comment]
Thanks @samfrjn11 for the detailed steps! Now I can reproduce this, however it happens only when there is `ListStyle` plugin added to the editor. Indeed it was easier to reproduce with two words in the last list item, but I managed to reproduced it somehow with one word too, you just need to be sure that triple-clicking selects the line till its very end.  
Here's the screencast:  

https://user-images.githubusercontent.com/34380544/107486304-2aee1300-6b85-11eb-898a-af76b154eb39.mp4
  
## To sum up:
1. Create editor with `ListStyle` plugin
2. Create the following structure:
```plaintext
heading
    - list item 1
    - list item 2
heading
    - list item 3
```
3. Select whole line with `list item 2 `
4. Press `delete`

Error:
```js
Uncaught TypeError: Cannot read property 'is' of null
    at Function._createAt (position.js:931)
    at Bl.createPositionAt (model.js:635)
    at ol.createPositionAt (writer.js:607)
    at liststyleediting.js:194
    at Bl.change (model.js:178)
    at Bl.listenTo.priority (liststyleediting.js:178)
    at Bl.fire (emittermixin.js:221)
    at Bl.<computed> [as deleteContent] (observablemixin.js:259)
    at Object.callback (deletecommand.js:122)
    at Bl._runPendingChanges (model.js:830)
```
  
I'll change the name of the issue to better reflect the error.

**Comment by pomek** [comment]
While implementing the list style feature, there were many cases to handle. I am pretty sure that I didn't predict everything and the issue confirms that. 

There is an invalid block of the code: https://github.com/ckeditor/ckeditor5/blob/a127c971ba23e4e6ed58d83327db617d85063b24/packages/ckeditor5-list/src/liststyleediting.js#L172-L203

And the fixed version below:

```js
		// If so, update the `listStyle` attribute for the second list.
		this.listenTo( model, 'deleteContent', () => {
			if ( !firstMostOuterItem ) {
				return;
			}

			const firstMostOuterListStyle = firstMostOuterItem.getAttribute( 'listStyle' );
			firstMostOuterItem = null;

			// Find the first most-outer item list in the merged list.
			// A case when the first list item in the second list was merged into the last item in the first list.
			//
			// <listItem listIndent="0" listType="bulleted" listStyle="square">UL List item 1</listItem>
			// <listItem listIndent="0" listType="bulleted" listStyle="square">UL List item 2</listItem>
			// <listItem listIndent="0" listType="bulleted" listStyle="circle">[]UL List item 1</listItem>
			// <listItem listIndent="0" listType="bulleted" listStyle="circle">UL List item 2</listItem>
			const secondListMostOuterItem = getSiblingListItem( firstMostOuterItem.nextSibling, {
				sameIndent: true,
				listIndent: firstMostOuterItem.getAttribute( 'listIndent' ),
				direction: 'forward'
			} );

			// The list item cannot be specified if a non-list item is placed between two lists. Abort then. See: #8651
			if ( !secondListMostOuterItem ) {
				return;
			}

			model.change( writer => {
				const items = [
					secondListMostOuterItem,
					...getSiblingNodes( writer.createPositionAt( secondListMostOuterItem, 0 ), 'forward' )
				];

				for ( const listItem of items ) {
					writer.setAttribute( 'listStyle', firstMostOuterListStyle, listItem );
				}
			} );
		}, { priority: 'low' } );
```

We're still missing tests to cover the bug.

**Comment by Reinmar** [comment]
Still reproducible after latest changes in the list properties plugin.

**Comment by maxbarnas** [comment]
Unfortunately the fix proposed by @pomek does not work anymore. The problem requires more research and/or changes in the current code.

**Comment by pomek** [comment]
When you apply the specified diff, the error should not be displayed anymore:

```diff
diff --git a/packages/ckeditor5-list/src/listpropertiesediting.js b/packages/ckeditor5-list/src/listpropertiesediting.js
index 0d52b6006e..64e8accd7a 100644
--- a/packages/ckeditor5-list/src/listpropertiesediting.js
+++ b/packages/ckeditor5-list/src/listpropertiesediting.js
@@ -214,6 +214,11 @@ export default class ListPropertiesEditing extends Plugin {
 					direction: 'forward'
 				} );
 
+				if ( !secondListMostOuterItem ) {
+					firstMostOuterItem = null;
+					return;
+				}
+
 				const items = [
 					secondListMostOuterItem,
 					...getSiblingNodes( writer.createPositionAt( secondListMostOuterItem, 0 ), 'forward' )
```

In short, a mechanism was looking for a second list. Unfortunately, it can't be found because the `<heading>` element is between `<listItem>s`. Hence, we should escape and do not process anything.

**Comment by pomek** [comment]
=> https://github.com/ckeditor/ckeditor5/pull/11123.
